File: micropython/bus_display_led/leds/led_controller.py
# led_controller.py
import time
from machine import Pin
import urequests as requests
from bus_display_led.config import server_url, ledstrip_endpoint
from bus_display_led.wifi.wifi_sta import wifi_status


# Pin definitions
CLOCK_PIN = 18   # GPIO 18 for clock
DATA_PIN = 15    # GPIO 15 for data
LATCH_PIN = 5    # GPIO 2 for the latch pin (RCLK)
RESET_PIN = 19    # GPIO 19 for the register reset pin
OE_PIN = 2    # GPIO 35 for the register reset pin

# Mapping LEDs to register values
led_to_register = {
    1: 0b0100000000000000,
    2: 0b0010000000000000,
    3: 0b0001000000000000,
    4: 0b0000100000000000,
    5: 0b0000010000000000,
    6: 0b0000001000000000,
    7: 0b0000000000000010,
    8: 0b0000000000000100,
    9: 0b0000000000001000,
    10: 0b0000000000010000,
    11: 0b0000000000100000,
    12: 0b0000000001000000,
}

# Pin initialization
clock   = Pin(CLOCK_PIN, Pin.OUT)
data    = Pin(DATA_PIN, Pin.OUT)
latch   = Pin(LATCH_PIN, Pin.OUT)
reset   = Pin(RESET_PIN, Pin.OUT)
oe      = Pin(OE_PIN, Pin.OUT)

def pulse(pin):
    """
    Generates a clock pulse on the given pin. (+- 10kHz, max 15MHz)
    """
    time.sleep_us(25)
    pin.on()
    time.sleep_us(50)
    pin.off()
    time.sleep_us(25)


def feed_register(leds):
    """
    Sends a 16-bit value to the shift register, LSB first.

    :param row: The row number (1-4).
    :param leds: A 16-bit integer representing the LED states.
    """
    # Send each bit (from LSB to MSB)
    for i in range(16):
        bit = (leds >> i) & 1  # Extract the bit (starting from LSB)
        data.value(bit)       # Set the data line
        pulse(clock)     # Pulse the clock pin
    data.value(0)             # reset to 0 in idle state


def init_pins():
    print("Initialising pins")
    oe.on() #disable output (high impedance)
    latch.off()
    clock.off()
    data.off() 
    pulse(reset) # put low to reset shift register 
    reset.on() # keep on to prevent reseting the registers
    pulse(latch) # put 0s in latches
    oe.off() # keep down to enable output 

def update_leds(value):
    """ Update the LEDs by sending a value to the shift register."""
    feed_register(led_to_register[value])
    pulse(latch)  # Latch the data into the output register


# Fetch data and update LEDs
def fetch_update_leds():
    """
    Fetch LED strip status and update shift registers.
    """
    try:
        global last_response
        url = f"{server_url}{ledstrip_endpoint}{wifi_status['mac']}"
        last_response = requests.get(url, timeout=5)
        if last_response.status_code != 200:
            print("Error:", last_response.json())
            return last_response.status_code

        data = last_response.json()
        strips = data.get('strips', [])
        for strip in strips:
            leds = strip.get('v', [])
            register_value = 0
            for i, val in enumerate(leds):
                if val and (i + 1) in led_to_register:
                    register_value |= led_to_register[i + 1]
            feed_register(register_value)
        pulse(latch)
        return last_response.status_code
    
    except Exception as e:
        print("Exception:", e)
        return -1

File: cpp/bus_display_led/main/led_controller.cpp
// led_controller.cpp
#include "led_controller.h"
#include "rom/ets_sys.h"

const char* LEDController::TAG = "LED_CTRL";

// LED to register bit mapping (matching your MicroPython code)
const uint16_t LEDController::led_to_register[13] = {
    0x0000,                  // LED 0 (not used)
    0b0100000000000000,      // LED 1
    0b0010000000000000,      // LED 2
    0b0001000000000000,      // LED 3
    0b0000100000000000,      // LED 4
    0b0000010000000000,      // LED 5
    0b0000001000000000,      // LED 6
    0b0000000000000010,      // LED 7
    0b0000000000000100,      // LED 8
    0b0000000000001000,      // LED 9
    0b0000000000010000,      // LED 10
    0b0000000000100000,      // LED 11
    0b0000000001000000,      // LED 12
};

LEDController::LEDController() : initialized_(false) {
}

LEDController::~LEDController() {
    if (initialized_) {
        clear_all();
        gpio_reset_pin(CLOCK_PIN);
        gpio_reset_pin(DATA_PIN);
        gpio_reset_pin(LATCH_PIN);
        gpio_reset_pin(RESET_PIN);
        gpio_reset_pin(OE_PIN);
    }
}

bool LEDController::initialize() {
    ESP_LOGI(TAG, "Initializing LED controller pins");
    
    // Configure all pins as outputs
    gpio_config_t io_conf = {};
    io_conf.intr_type = GPIO_INTR_DISABLE;
    io_conf.mode = GPIO_MODE_OUTPUT;
    io_conf.pin_bit_mask = (1ULL << CLOCK_PIN) | (1ULL << DATA_PIN) | 
                           (1ULL << LATCH_PIN) | (1ULL << RESET_PIN) | (1ULL << OE_PIN);
    io_conf.pull_down_en = GPIO_PULLDOWN_DISABLE;
    io_conf.pull_up_en = GPIO_PULLUP_DISABLE;
    
    esp_err_t ret = gpio_config(&io_conf);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "GPIO config failed: %s", esp_err_to_name(ret));
        return false;
    }
    
    // Initialize pins to safe state (matching your MicroPython init_pins())
    gpio_set_level(OE_PIN, 1);        // Disable output (high impedance)
    gpio_set_level(LATCH_PIN, 0);
    gpio_set_level(CLOCK_PIN, 0);
    gpio_set_level(DATA_PIN, 0);
    
    // Reset shift register
    pulse_pin(RESET_PIN);             // Put low to reset shift register
    gpio_set_level(RESET_PIN, 1);     // Keep high to prevent resetting
    
    // Put 0s in latches
    pulse_pin(LATCH_PIN);
    
    // Enable output
    gpio_set_level(OE_PIN, 0);
    
    initialized_ = true;
    ESP_LOGI(TAG, "LED controller initialized successfully");
    return true;
}

void LEDController::pulse_pin(gpio_num_t pin) {
    ets_delay_us(PULSE_DELAY_US);
    gpio_set_level(pin, 1);
    ets_delay_us(PULSE_DELAY_US * 2);
    gpio_set_level(pin, 0);
    ets_delay_us(PULSE_DELAY_US);
}

void LEDController::feed_register(uint16_t value) {
    // Send each bit (from LSB to MSB) - matching your MicroPython code
    for (int i = 0; i < 16; i++) {
        int bit = (value >> i) & 1;  // Extract the bit (starting from LSB)
        gpio_set_level(DATA_PIN, bit);
        pulse_pin(CLOCK_PIN);
    }
    gpio_set_level(DATA_PIN, 0);     // Reset to 0 in idle state
}

void LEDController::latch_data() {
    pulse_pin(LATCH_PIN);
}

void LEDController::set_leds(uint16_t pattern) {
    if (!initialized_) {
        ESP_LOGE(TAG, "LED controller not initialized");
        return;
    }
    
    feed_register(pattern);
    latch_data();
    ESP_LOGD(TAG, "Set LEDs with pattern: 0x%04X", pattern);
}

void LEDController::set_single_led(int led_number) {
    if (!initialized_) {
        ESP_LOGE(TAG, "LED controller not initialized");
        return;
    }
    
    if (led_number < 1 || led_number > 12) {
        ESP_LOGW(TAG, "Invalid LED number: %d (valid range: 1-12)", led_number);
        return;
    }
    
    uint16_t pattern = led_to_register[led_number];
    set_leds(pattern);
    ESP_LOGI(TAG, "Set LED %d (pattern: 0x%04X)", led_number, pattern);
}

void LEDController::clear_all() {
    if (!initialized_) {
        ESP_LOGE(TAG, "LED controller not initialized");
        return;
    }
    
    set_leds(0x0000);
    ESP_LOGI(TAG, "All LEDs cleared");
}

void LEDController::test_pattern() {
    if (!initialized_) {
        ESP_LOGE(TAG, "LED controller not initialized");
        return;
    }
    
    // Create pattern for LEDs 1, 3, 5, 7, 9, 11
    uint16_t pattern = led_to_register[1] | led_to_register[3] | led_to_register[5] | 
                       led_to_register[7] | led_to_register[9] | led_to_register[11];
    
    set_leds(pattern);
    ESP_LOGI(TAG, "Test pattern set: LEDs 1,3,5,7,9,11 (pattern: 0x%04X)", pattern);
}

void LEDController::set_from_array(const bool states[], size_t count) {
    uint16_t pattern = 0;
    for (size_t i = 0; i < count && i < 13; i++) {
        if (states[i]) {
            pattern |= led_to_register[i];
        }
    }
    set_leds(pattern);
}